### ISSUE TITLE
Not getting packet events

### ISSUE THREAD

**Issue by ayan4m1** [JRakNet/JRakNet#78]
I am expecting to be able to send simple packets between client and server. I am likely missing something about the design or implmentation of the library, but I'm stuck so I thought I'd open an issue.

Pastebin containing reproduction source: https://pastebin.com/n50dq6QR

Output log of code linked above:
````
Connected to the target VM, address: '127.0.0.1:64811', transport: 'socket'
17:50:20.541 [main] DEBUG io.netty.util.internal.logging.InternalLoggerFactory - Using SLF4J as the default logging framework
17:50:20.571 [main] DEBUG io.netty.channel.MultithreadEventLoopGroup - -Dio.netty.eventLoopThreads: 8
17:50:20.629 [main] DEBUG io.netty.util.internal.PlatformDependent - Platform: Windows
17:50:20.635 [main] DEBUG io.netty.util.internal.PlatformDependent0 - -Dio.netty.noUnsafe: false
17:50:20.636 [main] DEBUG io.netty.util.internal.PlatformDependent0 - Java version: 8
17:50:20.639 [main] DEBUG io.netty.util.internal.PlatformDependent0 - sun.misc.Unsafe.theUnsafe: available
17:50:20.641 [main] DEBUG io.netty.util.internal.PlatformDependent0 - sun.misc.Unsafe.copyMemory: available
17:50:20.642 [main] DEBUG io.netty.util.internal.PlatformDependent0 - java.nio.Buffer.address: available
17:50:20.643 [main] DEBUG io.netty.util.internal.PlatformDependent0 - direct buffer constructor: available
17:50:20.645 [main] DEBUG io.netty.util.internal.PlatformDependent0 - java.nio.Bits.unaligned: available, true
17:50:20.645 [main] DEBUG io.netty.util.internal.PlatformDependent0 - jdk.internal.misc.Unsafe.allocateUninitializedArray(int): unavailable prior to Java9
17:50:20.645 [main] DEBUG io.netty.util.internal.PlatformDependent0 - java.nio.DirectByteBuffer.<init>(long, int): available
17:50:20.646 [main] DEBUG io.netty.util.internal.PlatformDependent - sun.misc.Unsafe: available
17:50:20.647 [main] DEBUG io.netty.util.internal.PlatformDependent - -Dio.netty.tmpdir: C:\Users\ANDREW~1\AppData\Local\Temp (java.io.tmpdir)
17:50:20.648 [main] DEBUG io.netty.util.internal.PlatformDependent - -Dio.netty.bitMode: 64 (sun.arch.data.model)
17:50:20.651 [main] DEBUG io.netty.util.internal.PlatformDependent - -Dio.netty.noPreferDirect: false
17:50:20.652 [main] DEBUG io.netty.util.internal.PlatformDependent - -Dio.netty.maxDirectMemory: 1883242496 bytes
17:50:20.652 [main] DEBUG io.netty.util.internal.PlatformDependent - -Dio.netty.uninitializedArrayAllocationThreshold: -1
17:50:20.654 [main] DEBUG io.netty.util.internal.CleanerJava6 - java.nio.ByteBuffer.cleaner(): available
17:50:20.738 [main] DEBUG io.netty.channel.nio.NioEventLoop - -Dio.netty.noKeySetOptimization: false
17:50:20.738 [main] DEBUG io.netty.channel.nio.NioEventLoop - -Dio.netty.selectorAutoRebuildThreshold: 512
17:50:20.764 [main] DEBUG io.netty.util.internal.PlatformDependent - org.jctools-core.MpscChunkedArrayQueue: available
17:50:20.832 [Thread-1] INFO  com.whirvis.jraknet.server.RakNetServer - Added listener in.thekreml.MainTest$1
17:50:20.890 [Thread-1] DEBUG io.netty.channel.DefaultChannelId - -Dio.netty.processId: 11572 (auto-detected)
17:50:20.895 [Thread-1] DEBUG io.netty.util.NetUtil - -Djava.net.preferIPv4Stack: false
17:50:20.895 [Thread-1] DEBUG io.netty.util.NetUtil - -Djava.net.preferIPv6Addresses: false
17:50:21.110 [Thread-1] DEBUG io.netty.util.NetUtil - Loopback interface: lo (Software Loopback Interface 1, 127.0.0.1)
17:50:21.112 [Thread-1] DEBUG io.netty.util.NetUtil - Failed to get SOMAXCONN from sysctl and file \proc\sys\net\core\somaxconn. Default: 200
17:50:21.297 [Thread-1] DEBUG io.netty.channel.DefaultChannelId - -Dio.netty.machineId: 00:15:5d:ff:fe:3b:1b:26 (auto-detected)
17:50:21.307 [Thread-1] DEBUG io.netty.util.internal.InternalThreadLocalMap - -Dio.netty.threadLocalMap.stringBuilder.initialSize: 1024
17:50:21.308 [Thread-1] DEBUG io.netty.util.internal.InternalThreadLocalMap - -Dio.netty.threadLocalMap.stringBuilder.maxSize: 4096
17:50:21.327 [Thread-1] DEBUG io.netty.util.ResourceLeakDetector - -Dio.netty.leakDetection.level: simple
17:50:21.327 [Thread-1] DEBUG io.netty.util.ResourceLeakDetector - -Dio.netty.leakDetection.maxRecords: 4
17:50:21.327 [Thread-1] DEBUG io.netty.util.ResourceLeakDetector - -Dio.netty.leakDetection.maxSampledRecords: 40
17:50:21.371 [Thread-1] DEBUG io.netty.buffer.PooledByteBufAllocator - -Dio.netty.allocator.numHeapArenas: 8
17:50:21.372 [Thread-1] DEBUG io.netty.buffer.PooledByteBufAllocator - -Dio.netty.allocator.numDirectArenas: 8
17:50:21.372 [Thread-1] DEBUG io.netty.buffer.PooledByteBufAllocator - -Dio.netty.allocator.pageSize: 8192
17:50:21.372 [Thread-1] DEBUG io.netty.buffer.PooledByteBufAllocator - -Dio.netty.allocator.maxOrder: 11
17:50:21.372 [Thread-1] DEBUG io.netty.buffer.PooledByteBufAllocator - -Dio.netty.allocator.chunkSize: 16777216
17:50:21.372 [Thread-1] DEBUG io.netty.buffer.PooledByteBufAllocator - -Dio.netty.allocator.tinyCacheSize: 512
17:50:21.372 [Thread-1] DEBUG io.netty.buffer.PooledByteBufAllocator - -Dio.netty.allocator.smallCacheSize: 256
17:50:21.373 [Thread-1] DEBUG io.netty.buffer.PooledByteBufAllocator - -Dio.netty.allocator.normalCacheSize: 64
17:50:21.373 [Thread-1] DEBUG io.netty.buffer.PooledByteBufAllocator - -Dio.netty.allocator.maxCachedBufferCapacity: 32768
17:50:21.373 [Thread-1] DEBUG io.netty.buffer.PooledByteBufAllocator - -Dio.netty.allocator.cacheTrimInterval: 8192
17:50:21.373 [Thread-1] DEBUG io.netty.buffer.PooledByteBufAllocator - -Dio.netty.allocator.useCacheForAllThreads: true
17:50:21.388 [Thread-1] DEBUG io.netty.buffer.ByteBufUtil - -Dio.netty.allocator.type: pooled
17:50:21.389 [Thread-1] DEBUG io.netty.buffer.ByteBufUtil - -Dio.netty.threadLocalDirectBufferSize: 65536
17:50:21.389 [Thread-1] DEBUG io.netty.buffer.ByteBufUtil - -Dio.netty.maxThreadLocalCharBufferSize: 16384
17:50:21.403 [Thread-1] INFO  com.whirvis.jraknet.server.RakNetServer - Started server
17:50:21.404 [Thread-1] INFO  in.thekreml.MainTest - Server started up.
17:50:21.585 [main] INFO  com.whirvis.jraknet.client.RakNetClient - Added listener in.thekreml.MainTest$2
17:50:21.586 [DiscoveryThread] INFO  com.whirvis.jraknet.client.discovery.DiscoveryThread - Started discovery thread
17:50:21.603 [main] DEBUG io.netty.buffer.AbstractByteBuf - -Dio.netty.buffer.bytebuf.checkAccessible: true
17:50:21.606 [main] DEBUG io.netty.util.ResourceLeakDetectorFactory - Loaded default ResourceLeakDetector: io.netty.util.ResourceLeakDetector@6239aba6
17:50:21.617 [main] DEBUG io.netty.util.Recycler - -Dio.netty.recycler.maxCapacityPerThread: 32768
17:50:21.617 [main] DEBUG io.netty.util.Recycler - -Dio.netty.recycler.maxSharedCapacityFactor: 2
17:50:21.617 [main] DEBUG io.netty.util.Recycler - -Dio.netty.recycler.linkCapacity: 16
17:50:21.618 [main] DEBUG io.netty.util.Recycler - -Dio.netty.recycler.ratio: 8
17:50:22.622 [main] INFO  com.whirvis.jraknet.client.SessionPreparation - session planner #-4501988052636712809 created server session using globally unique ID -5529632743147421254 and maximum transfer unit with size of 1492 bytes (11936 bits) for server address /127.0.0.1:55055
17:50:22.634 [nioEventLoopGroup-3-1] INFO  in.thekreml.MainTest - Client connected from 127.0.0.1
17:50:22.636 [nioEventLoopGroup-2-1] INFO  in.thekreml.MainTest - Client connected to 127.0.0.1
````

Specifically, I'd expect to see "Got message" or "Unknown message from" in the logs and I do not. Is there anything I am missing with regard to setting up bidirectional communication between server and client?

**Comment by whirvis** [comment]
The reason you are not getting packet events sometimes is because the call to send a packet is not being sent when the client is connected. If the client is not connected, it will refuse to send the message and return null (rather than the EncapsulatedPacket copy which can be used for various things, however most people don't need that other than to check if a packet was sent.) This can be fixed by putting the sendMessage() call in the onConnected loop, as that is called after the client has been connected to the server and is ready to send and handle packets.

The reason an IndexOutOfBoundsException is being thrown is because you are sending a packet with no data. As a result, the RakNetPacket does not know what to do as a result and ends up throwing an IllegalArgumentException. This triggers a onHandlerException() method call which is called when an exception caused by a client is occurred. Since this method is a default method in the interface, I originally made it so that it would print out the caught exception. I have undone this and made it so only the onThreadException() prints out a stack trace by default since the onHandlerException() is caused by outside sources which could be malicious.

**TL;DR** You sent bad data on the client (write an unsigned byte to the packet so it has an ID for the server to read), and the server was printing an error report unnecessarily.

Thanks for the report! Once you update to 2.9.4 and fix the error in your code, let me know if you have any more problems :) Also, just to be clear, you were the person that E-mailed me earlier today right?

**Comment by whirvis** [comment]
Fixed in 2.9.4

**Comment by ayan4m1** [comment]
Yes, I had emailed you recently. Thank you very much for your time and response. I moved the packet send to onConnect() - I am still learning how to construct the packets but I am definitely getting further.

Great library so far, thanks again!

**Comment by whirvis** [comment]
No problem! What When in doubt, just reference the JavaDocs :) Ive documented nearly everything in the source code, so most classes/constructors/methods will tell you what they do and how their parameters should be used.

If you have any questions though, feel free to open an issue or send me an E-Mail.
